### PR TITLE
CLDC-2322 read and sign data sharing agreement

### PR DESCRIPTION
**Pull request by ghost**
https://digital.dclg.gov.uk/jira/browse/CLDC-2322

- Add dynamic data sharing agreement template
- Update details page to include a row for the data sharing agreement
- Update notification banner to dynamically handle a text paragraph
- persist DPO and org data on table to prevent future updates

### CODE REVIEW

File: app/helpers/data_sharing_agreement_helper.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    end
+  end
+
+  def name_for_data_sharing_agreement(data_sharing_agreement, user)
+    if data_sharing_agreement.present?
+      data_sharing_agreement.data_protection_officer.name
+    elsif user.is_dpo?
+      user.name
+    else
+      "[DPO name]"
```
Comment (natdeanlewissoftwire) [inline app/helpers/data_sharing_agreement_helper.rb:30]: don't we want this to be the actual name of the DPO?

File: app/views/organisations/data_sharing_agreement.html.erb [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+      <% if @data_sharing_agreement %>
+        This agreement is made the <%= @data_sharing_agreement.signed_at.day.ordinalize %> day of <%= @data_sharing_agreement.signed_at.strftime("%B") %> <%= @data_sharing_agreement.signed_at.year %>
+      <% elsif current_user.is_dpo? %>
+        This agreement is made the <%= Time.zone.now.day.ordinalize %> day of <%= Time.zone.now.strftime("%B") %> <%= Time.zone.now.year %>
+      <% else %>
+        This agreement is made the [XX] day of [XX] 20[XX]
+      <% end %>
+    </p>
+    <p class="govuk-body-m"><strong>between</strong></p>
+    <p class="govuk-body-m">1) [core data provider organisation] of [full address] (“CORE Data Provider”)</p>
```
Comment (natdeanlewissoftwire) [inline app/views/organisations/data_sharing_agreement.html.erb]: shouldn't these [] be populated with actual data?

**Review by natdeanlewissoftwire — COMMENTED** [review]
have done a first pass, looks good - have a feeling my queries are just me misunderstanding the ticket but I've left them in anyway

File: app/helpers/data_sharing_agreement_helper.rb (hunk 1 above)
Comment (ghost) [reply app/helpers/data_sharing_agreement_helper.rb:30]: yes but but only when not DPO or when not signed!

File: app/views/organisations/data_sharing_agreement.html.erb (hunk 2 above)
Comment (ghost) [reply app/views/organisations/data_sharing_agreement.html.erb]: apparently not, only certain bits in the document need to be replaced

**Review by natdeanlewissoftwire — APPROVED** [review]

File: app/views/organisations/data_sharing_agreement.html.erb (hunk 2 above)
Comment (ghost) [reply app/views/organisations/data_sharing_agreement.html.erb]: I was wrong, these need updating! Will fix shortly

File: app/views/organisations/data_sharing_agreement.html.erb [hunk 3]
```diff
@@ -15,7 +15,11 @@
       <% end %>
```
Comment (natdeanlewissoftwire) [inline app/views/organisations/data_sharing_agreement.html.erb:15]: maybe some of this logic should go in a helper

**Review by natdeanlewissoftwire — APPROVED** [review]
looks good, possible case for a helper but other than that nothing from me :)

**Pull request closed without merging**